Elsa Localization Support

mohdali opened the pull request:
This PR adds support for Localization in Elsa Studio by utilizing ASP.NET framework for localization.

This PR does not include any localization by itself but enables the users to provide their own localizers by supplying a `ILocalizationProvider` class with the required resource files.

The base Resource file is added to `Elsa.Studio.Core` project that contains all the strings found in Elsa Studio source code.

Localization of Activities is also considered. In that case, additional strings may need to be registered in the Resource file to match the activity name, inputs, outputs, etc.

This is an initial effort and there may be many areas for improvement. However, it may be a good start for more localization support.

https://github.com/user-attachments/assets/d9205c43-6c39-4143-9c5b-d0709bd388f0

Guidance for Localization (README included in PR).

# Elsa Studio Localization
Elsa Studio Localization Allows localizing the Studio UI to different languages.

`ILocalizationProvider` is the interface that needs to be implemented to provide the localized strings for the Studio UI.

Create your own localization file (Using the resource file or any alternative method' and implement the `ILocalizationProvider` interface.

Afterwards, register your implementation in the DI container.

```cssharp
builder.Services.AddSingleton<ILocalizationProvider, MyLocalizationProvider>();
```

## Steps to Enable Localization

### For Blazor Server
1. Add Reference to `Elsa.Studio.Localization.BlazorServer` package.
1. Define the Localization Configuration
    ```csharp
    var localizationConfig = new LocalizationConfig
    {
        ConfigureLocalizationOptions = options => configuration.GetSection("Localization").Bind(options),
    };
    ```
1. Register the Localization Module `builder.Services.AddLocalizationModule();`
1. Register your locallization provider
	```csharp
	builder.Services.AddSingleton<ILocalizationProvider, MyLocalizationProvider>();
	```
1. Add The localization Middleware. Making Sure that Controllers are also mapped.
    ```csharp
    app.UseElsaLocalization();
    app.MapControllers();
    ```
1. Add below configuration in the `appsettings.json` file, specifying the supported cultures.
    ```json
    "Localization": {
        "SupportedCultures": [
          "ja-JP",
          "fr-FR"
        ]
      }
    ```

### For Blazor WebAssembly
1. Add Reference to `Elsa.Studio.Localization.BlazorWasm` package.
1. Define the Localization Configuration
    ```csharp
    var localizationConfig = new LocalizationConfig
    {
        ConfigureLocalizationOptions = options => configuration.GetSection("Localization").Bind(options),
    };
    ```
1. Register the Localization Module `builder.Services.AddLocalizationModule();`
1. Register your locallization provider
	```csharp
	builder.Services.AddSingleton<ILocalizationProvider, MyLocalizationProvider>();
	```
1. Use the localization Middleware
    ```csharp
    await app.UseElsaLocalization();
    ```
1. Add below configuration in the `appsettings.json` file, specifying the supported cultures.
    ```json
    "Localization": {
        "SupportedCultures": [
          "ja-JP",
          "fr-FR"
        ]
      }
    ```

Soumyas38 committed:
Localization for elsa studio.

Soumyas38 committed:
Latest Changes.

Commit:
WIP Localization Module

Commit:
cleanup

Commit:
BlazorServer implementation

Soumyas38 committed:
Elsa studio changes for Localization.

Commit:
Update localization options

Commit:
merge from soumya

Commit:
refactoring

Soumyas38 committed:
Added translation for components.

Soumyas38 committed:
Deleted Unused resource files.

Commit:
Merge remote-tracking branch 'soumya/Soumya_Studio' into elsa-localization

Commit:
Fix build

Soumyas38 committed:
Removed extra localization files and added missing translations.

Commit:
add controllers

Commit:
merge soumya

Commit:
Fix Imports

Commit:
Reconfigure Localization Service with the ResourceManager

Commit:
Add Blazor Wasm Localization

Commit:
Refactor LocalizationService

Commit:
Rename Localization Service

Soumyas38 committed:
Translation added for Agents and Secrets.

Commit:
refactoring

Commit:
Merge remote-tracking branch 'soumya/Soumya_Studio' into elsa-localization

Commit:
merge from soumya

Commit:
Refactor localization

Commit:
cleanup resources

Commit:
additional cleanup

Commit:
Delete Japanese localization

Commit:
clean up

Commit:
remove specific functions

Commit:
Add a README for localization

Commit:
Resolve merge conflicts

Commit:
Update localization

Commit:
Fix merge errors

Commit:
fix formatting

Commit:
Use Timestamp component

sfmskywalker committed:
Merge branch 'main' into elsa-localization

sfmskywalker commented on src/framework/Elsa.Studio.Shared/Components/DataPanel.razor:
```diff
@@ -21,11 +24,18 @@
+                        }
+                        else
+                        {
+                            @if (@item.Key == "Created" || @item.Key == "Updated" || item.Key == "Finished")
```
Although I like that we would have better display support for Timestamp values, I don't like that this DataPanel component has specific logic for these properties. Either we should revert this change, or, even better, we make DataPanel support components as values in addition to string values.

sfmskywalker replied:
Will fix later

sfmskywalker reviewed (commented):
There are some small changes that I want to take a little bit more time for in order to understand them, but overall, this is an amazing job, thank you for this awesome contribution @mohdali 🔥

sfmskywalker committed:
Implement ILocalizer interface and replace ElsaLocalization

Created a new ILocalizer interface for localization purposes. Replaced ElsaLocalization with ILocalizer throughout the codebase for consistency and better readability. Removed unnecessary using statements and simplified namespaces.

sfmskywalker committed:
Remove unused using directives across the project

Cleared redundant using directives from ILocalizationProvider, DefaultLocalizer, and ActivityWrapper components. This cleanup improves code readability and may slightly enhance compilation performance.

sfmskywalker committed:
Refactor constructors to use parameterized field assignment

Converted multiple class constructors to use parameterized field assignment directly in the constructor. This change reduces boilerplate code and improves readability. Additionally, some unnecessary using directives were removed.

sfmskywalker committed:
Add XML documentation to DashboardMenu class

Added summary XML documentation to the DashboardMenu class and the GetMenuItemsAsync method to enhance code readability and maintainability. This documentation provides a brief description of the class's purpose and the method's functionality.

sfmskywalker committed:
Update language picker to use customizable formatter

This change updates the `LanguagePicker` component to utilize a customizable formatter defined in `LocalizationOptions`. It allows for more flexibility in how culture names are displayed in the UI, enhancing localization support. The default formatter is set to use the culture's display name.

sfmskywalker committed:
Update culture picker to use culture name

Changed the CulturePickerFormatter to use the culture's Name property instead of DisplayName. This ensures consistency in the format used for culture identifiers.

sfmskywalker committed:
Merge branch 'main' into elsa-localization
